File: 백준_C++/2146.cpp
#include<bits/stdc++.h>
using namespace std;

int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};
int N, board[101][101], dist[101][101], land[101][101];
queue<pair<int, int>> Q;
int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> N;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++) 
            cin >> board[i][j];

    
    int mapNo = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (board[i][j] == 0 || land[i][j] != 0) continue;
            land[i][j] = ++mapNo;
            Q.push({i, j});

            while(!Q.empty()) {
                auto cur = Q.front(); Q.pop();
                for (int dir = 0; dir < 4; dir++) {
                    int nx = dx[dir] + cur.first;
                    int ny = dy[dir] + cur.second;
                    if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
                    if (board[nx][ny] == 0 || land[nx][ny] != 0) continue;
                    land[nx][ny] = mapNo;
                    Q.push({nx, ny});
                }
            }
        }
    }

    for (int i = 0; i < N; i++)
        fill(dist[i], dist[i] + N, -1);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (land[i][j] != 0) {
                Q.push({i, j});
                dist[i][j] = 0;
            }
        }
    }

    int ans = 999999;

    while (!Q.empty()) {
        auto cur = Q.front(); Q.pop();
        for (int dir = 0; dir < 4; dir++) {
            int nx = dx[dir] + cur.first;
            int ny = dy[dir] + cur.second;
            if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;

            if (land[nx][ny] != 0 && land[nx][ny] != land[cur.first][cur.second]) {
                ans = min(ans, dist[cur.first][cur.second] + dist[nx][ny]);
            }

            if (board[nx][ny] == 0 && dist[nx][ny] == -1) {
                land[nx][ny] = land[cur.first][cur.second];
                dist[nx][ny] = dist[cur.first][cur.second] + 1;
                Q.push({nx, ny});
            }
        }
    }

    cout << ans;
}

File: 백준_Java/2146.java
import java.io.*;
import java.util.*;

class Main {
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};
    static int n;
    static int[][] board;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        board = new int[n][n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.print(solution());
    }

    static int solution() {
        Queue<Pair> q = new LinkedList<>();
        int[][] dist = new int[n][n];
        int[][] land = new int[n][n];
        int answer = 1000000;
        int mapNo = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 0 || land[i][j] != 0) continue;
                land[i][j] = ++mapNo;
                q.add(new Pair(i, j));

                while (!q.isEmpty()) {
                    Pair cur = q.poll();
                    for (int dir = 0; dir < 4; dir++) {
                        int nx = cur.x + dx[dir];
                        int ny = cur.y + dy[dir];
                        if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
                        if (board[nx][ny] == 0 || land[nx][ny] != 0) continue;
                        land[nx][ny] = mapNo;
                        q.add(new Pair(nx, ny));
                    }
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (land[i][j] != 0) {
                    q.add(new Pair(i, j));
                    dist[i][j] = 1;
                }
            }
        }

        while (!q.isEmpty()) {
            Pair cur = q.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur.x + dx[dir];
                int ny = cur.y + dy[dir];
                if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
                
                if (land[nx][ny] != 0 && land[nx][ny] != land[cur.x][cur.y]) {
                    answer = Math.min(answer, dist[cur.x][cur.y] + dist[nx][ny]);
                }

                if (dist[nx][ny] > 0 || board[nx][ny] == 1) continue;
                land[nx][ny] = land[cur.x][cur.y];
                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                q.add(new Pair(nx, ny));
            }
        }

        return answer - 2;
    }

    static class Pair {
        int x;
        int y;
        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}